Update bpfd README

anfredette opened the pull request:
Better describe what bpfd is and the problems it solves.

Addresses issue #474.

dave-tucker commented on README.md:
```diff
@@ -1,65 +1,112 @@
+
+### What is bpfd?
+
+bpfd is a system daemon that supports, loading, unloading, modifying, and
+monitoring BPF programs. The solution for Kubernetes uses an operator, built
+with the Operator SDK, that can deploy the bpfd daemon and a bpfd agent to all
+the nodes in a cluster. The agent supports well-defined Kubernetes APIs (CRDs)
+that can be used to manage the lifecycle of BPF programs across a Kubernetes
+cluster.
```
```suggestion
bpfd is a software stack that aims to make it easy to load, unload, modify and monitor BPF programs whether on a single host, or in a Kubernetes cluster. It is comprised of:

- bpfd: A system daemon that supports load, unload, modifying and monitoring of BPF programs exposed over a gRPC API
- eBPF CRDS: That provides a way to express intent to load BPF programs `XdpProgram`, `TcProgram` etc... and represent the runtime state of loaded programs via `BpfProgram`.
- bpfd-agent: That ensures that the `BpfPrograms` for a given node are in the desired state
- bpfd-operator: An operator, built using [Operator SDK](https://sdk.operatorframework.io/) that manages the installation and lifecycle of bpfd-agent and the CRDs in a Kubernetes cluster.
```

dave-tucker commented on README.md:
```diff
@@ -1,65 +1,112 @@
+The benefits of this solution include the following:
+
+- Security
+  - Improved security because only the bpfd daemon, which can be tightly
```
You could mention here that only a single thread keeps these capabilities and the other threads (serving RPCs) do not.

dave-tucker commented on README.md:
```diff
@@ -1,65 +1,112 @@
+  - Simplifies the deployment of BPF programs in a Kubernetes cluster.
+  - Improved ease of use and productivity because developers can use a standard
+    framework for managing BPF programs rather than developing their own.
+  - Developers can still use Cilium/libxdp/Aya/etc libraries for BPF
```
s/libxdp/libbpf/g

anfredette replied:
I changed this, but
a) can't they use libxdp too?
b) I didn't do /g because the other instances were where I mentioned libxdp multiprog.

dave-tucker replied:
You could, but I don't feel its a good example:

libxdp is used for:
a. loading and attaching xdp programs
b. working with af_xdp sockets

if you move (a) to bpfd, then you aren't left with much

anfredette replied:
I'll leave libxdp out to keep things simple, but I feel like it's disingenuous for us to say you can't attach multiple programs to an XDP hookpoint without bpfd because you can -- you just need to use libxdp.

anfredette replied:
Never mind that comment I just made and deleted.  I agree with you.

dave-tucker reviewed (changes requested):
Overall a massive improvement, Thank You 🙌  Some changes required to the content.
I'd also run through a spell checker - I doubt I picked up everything.

anfredette commented on README.md:
```diff
@@ -1,65 +1,112 @@
+  - These permissions give the pods great powers.
+- Handling multiple BPF programs
+  - Not all libraries support multiple XDP programs sharing the same interface.
+  - When multiple BPF programs are used, they can interfere with each other in
+    unpredictable ways.
```
I agree with your addition, but think the existing bullet is valid too.

dave-tucker replied:
Maybe? I think it needs some additional qualification because as it stands it could mean that an XDP program could interfere with a Tracepoint program in unexpected ways.

anfredette replied:
I could qualify it as:
- they can interfere with each other in unpredictable ways
+ some BPF programs can interfere with others in unpredictable ways.

However, for the example given, if I have a tracepoint in some code that I'm expecting to be hit based on receiving a packet, and if someone else installs an XDP program that drops those packets, then I might wonder why my tracepoint isn't getting hit.  In that case, it might be good to know that some other app installed that XDP program and I might want to investigate it.

anfredette commented on README.md:
```diff
@@ -1,65 +1,112 @@
+    on what they need to do).
+  - These permissions give the pods great powers.
+- Handling multiple BPF programs
+  - Not all libraries support multiple XDP programs sharing the same interface.
```
Are there hooks other than XDP that aren't designed to support multiple programs?  Also, for XDP, this is only the case if you don't use libxdp or bpfd.

anfredette commented on README.md:
```diff
@@ -1,81 +1,148 @@
+  - The cluster administrator may not be aware that BPF programs are being used
+    in a cluster.
+  - It is possible for some BPF programs to interfere with others in
+    unpredictable ways.
```
@dave-tucker FYI, I generalized this comment and moved it down to the "debugging" bullet.

astoycos commented:
Hrm something's going on with netlify we're not getting automatic previews

astoycos commented on README.md:
````diff
@@ -1,81 +1,148 @@
-sudo ./target/debug/bpfd&
-sudo ./target/debug/bpfctl load --path /path/to/xdp/program --section-name "pass" xdp -i wlp2s0 --priority 50
-```
+bpfd is a system daemon for managing BPF programs that works both on a single
````
```suggestion
bpfd is a system daemon for managing BPF programs that works both on a single linux
```

astoycos commented on README.md:
```diff
@@ -1,81 +1,148 @@
+    damage (either unintentionally or intentionally).
+- Handling multiple BPF programs on the same BPF hooks
+  - Not all BPF hooks are designed to support multiple programs.
+  - Some software using BPF assumes exclusive use of a BPF hook and can
+    unintentionally eject existing programs when being attached.
```
```suggestion
  - Some software using BPF assumes exclusive use of a BPF hook and can
    unintentionally eject existing programs when being attached.  Which can result in silent failures and indeterministic behavior.
```

astoycos commented on README.md:
```diff
@@ -1,81 +1,148 @@
+	  - This requires careful design and intricate knowledge of the BPF program
+	    lifecycle to ensure your program stays loaded and that you can easily
+	    tolerate pod restarts and upgrades.
+
```
What about fine grain versioning problems? 

Something along the lines of....

- In Bpf enabled K8s deployments today the Bpf Program is often embedded into the userspace binary that loads and interacts with it.  This means there's no easy way to have fine-grained versioning control  of the bpfProgram in relation to it's accompanying userspace counterpart 

Or something like that :)

astoycos commented on README.md:
```diff
@@ -1,81 +1,148 @@
+## What is bpfd?
+
+bpfd is a software stack that aims to make it easy to load, unload, modify and
+monitor BPF programs whether on a single host, or in a Kubernetes cluster. bpfd comprises:
```
```suggestion
monitor BPF programs whether on a single host, or in a Kubernetes cluster. bpfd includes the following core components:
```

astoycos commented on README.md:
```diff
@@ -1,81 +1,148 @@
+    single thread keeps these capabilities while the other threads (serving
+    RPCs) do not.
+  - Gives the administrators control over who can load programs.
+  - Allows administrators to define rules for the ordering of networking BPF
```
road map?

anfredette replied:
Yes.  We plan to build policies around this, but it doesn't exist yet.

Billy99 commented on README.md:
```diff
@@ -1,81 +1,148 @@
+- There are no standard libraries for loading and managing BPF programs leaving
+  each application developer to implement their own.
+- Deployment on Kubernetes is not simple.
+	  - It is an involved process that requires first writing a daemon that loads
```
This isn't rendoring correctly

Billy99 commented on README.md:
```diff
@@ -1,81 +1,148 @@
+## What is bpfd?
+
+bpfd is a software stack that aims to make it easy to load, unload, modify and
+monitor BPF programs whether on a single host, or in a Kubernetes cluster. bpfd comprises:
```
Read a little funny. Maybe change to:
```suggestion
monitor BPF programs whether on a single host, or in a Kubernetes cluster. bpfd is comprised of:
```

anfredette replied:
Everyone is commenting on this :-).  Dave said what you said.  I changed his suggestion to what you see because that is the proper English usage (i.e., the whole comprises the parts).  However, many people use it as you and Dave suggest, so that is becoming more accepted, though is still controversial.  The non-controversial option is to say "is composed of".  However, I'm going with Andrew's suggestion: "includes the following core components".

astoycos commented on README.md:
```diff
@@ -1,81 +1,148 @@
+    of new programs atomic.
+- Productivity
+  - Simplifies the deployment of BPF programs in a Kubernetes cluster.
+  - Improved ease of use and productivity because developers can use a standard
```
```suggestion
  - Improved ease of use and productivity, developers can stop worrying about program lifecycle (loading, attaching, pin management, etc) and use existing bpf libraries to interact with their program maps using well defined pin points which are managed by bpfd.
```

astoycos commented on README.md:
```diff
@@ -1,81 +1,148 @@
+    framework for managing BPF programs rather than developing their own.
+  - Developers can still use Cilium/libbpf/Aya/etc libraries for BPF
+    development, and load/unload with bpfd.
+  - Provides BPF Bytecode Image Specifications that allows fine-grained separate
```
Ahh cool You had this here :)

You can ignore my comment above if you want

anfredette replied:
I think it's okay -- the first comment describes the problem, while this one describes the solution.

astoycos reviewed (commented):
Thanks for this @anfredette this is like the most important readme IMO :), Just had a few comments, feel free to take them with a grain of salt

Billy99 commented on README.md:
```diff
@@ -1,81 +1,148 @@
+
+  - bpfd: A system daemon that supports loading, unloading, modifying and
+	  monitoring of BPF programs exposed over a gRPC API.
+  - eBPF CRDS: That provide a way to express intent to load BPF programs
```
Not sure how to show updates across multiple lines, but consider rewording to:
```
- eBPF CRDS: bpfd provides a set of CRDs (`XdpProgram`, `TcProgram`, etc.) that provide a way to express intent to load BPF programs as well as a bpfd generated CRD ( `BpfProgram`) used to represent the runtime state of loaded programs.
```

Billy99 commented on README.md:
```diff
@@ -1,81 +1,148 @@
+
+## Challenges for BPF in Kubernetes
+
+- Requires privileged pods
```
```suggestion
- Requires privileged pods.
```

Billy99 commented on README.md:
```diff
@@ -1,81 +1,148 @@
+  - Since the Linux capabilities are very broad it is challenging to constrain a
+    pod to the minimum set of privileges required. This can allow them to do
+    damage (either unintentionally or intentionally).
+- Handling multiple BPF programs on the same BPF hooks
```
```suggestion
- Handling multiple BPF programs on the same BPF hooks.
```

Billy99 commented on README.md:
```diff
@@ -1,81 +1,148 @@
+  - eBPF CRDS: That provide a way to express intent to load BPF programs
+	  `XdpProgram`, `TcProgram`, etc. and represent the runtime state of loaded
+	  programs via `BpfProgram`. 
+  - bpfd-agent: That ensures that the `BpfPrograms` for a given node are in the
```
```suggestion
  - bpfd-agent: A container running in the bpfd daemonset that ensures that the requested BPF programs for a given node are in the
```

anfredette replied:
I tweaked your suggestion a little.

Billy99 commented on README.md:
```diff
@@ -1,81 +1,148 @@
+  - bpfd-agent: That ensures that the `BpfPrograms` for a given node are in the
+	  desired state.
+  - bpfd-operator: An operator, built using [Operator
+	  SDK](https://sdk.operatorframework.io/) that manages the installation and
```
```suggestion
	  SDK](https://sdk.operatorframework.io/), that manages the installation and
```

anfredette commented on README.md:
````diff
@@ -1,81 +1,148 @@
-sudo ./target/debug/bpfd&
-sudo ./target/debug/bpfctl load --path /path/to/xdp/program --section-name "pass" xdp -i wlp2s0 --priority 50
-```
+bpfd is a system daemon for managing BPF programs that works both on a single
+node and in a Kubernetes cluster using the bpfd operator and CRDs.
````
I think Linux is true today. Would it ever run on anything but Linux?  Windows?

anfredette commented on README.md:
```diff
@@ -1,81 +1,148 @@
+    unpredictable ways.
+  - SSH access or a privileged pod is necessary to determine the state of BPF
+    programs on each node in the cluster.
+- There are no standard libraries for loading and managing BPF programs leaving
+  each application developer to implement their own.
```
I agree that could be improved.  Your statement about k8s is a good one.  On non-k8s, aren't there still challenges with the "managing" part, or the lifecycle management?  And, I assume it's still the case that Cilium doesn't support TC.  I'll do an update on this.

anfredette replied:
Actually, the k8s issues are kind of covered in the next bullet.  I'll push an update and you can see what you think.

anfredette commented on README.md:
```diff
@@ -1,81 +1,148 @@
+    to allow multiple XDP programs on single interface
+  - This same protocol is also supported for TC programs to provide a common
+    multi-program user experience across both TC and XDP.
+  - An additional benefit for TC is that it makes the re-ordering and inserting
```
It is.  And, TC may change, so I'm just going to delete that bullet.

anfredette commented:
@dave-tucker @astoycos @Billy99 Thank you all for your careful reading and improvements!  It's important to get this right.  I think I've addressed all of your comments now, so please take another look when you have a chance.

astoycos commented:
+1 from me I'll let the other's approve + lgtm

Billy99 commented on README.md:
```diff
@@ -1,81 +1,160 @@
+
+  - bpfd: A system daemon that supports loading, unloading, modifying and
+	  monitoring of BPF programs exposed over a gRPC API.
+  - eBPF CRDS: That provide a way to express intent to load BPF programs
```
You left the original line in and then added my suggestion as a new bullet. Delete this original bullet item.

Billy99 commented:
LGTM

astoycos reviewed (approved):
LGTM
